OPHD/States/Planet.cpp

```diff
@@ -66,3 +68,118 @@ void Planet::update()
 	const auto spriteFrameOffset = NAS2D::Point{mTick % 8 * PlanetSize.x, ((mTick % 64) / 8) * PlanetSize.y};
 	renderer.drawSubImage(mImage, mPosition, NAS2D::Rectangle<int>::Create(spriteFrameOffset, PlanetSize));
 }
+
+namespace
+{
+	Planet::Attributes parsePlanet(const NAS2D::Xml::XmlElement* xmlNode);
+
+	void parseElementValue(Planet::PlanetType& destination, const NAS2D::Xml::XmlElement* element);
+	void parseElementValue(Planet::Hostility& destination, const NAS2D::Xml::XmlElement* element);
+
+	const std::unordered_map<std::string, Planet::PlanetType> planetTypeTable
+	{
+		{"None", Planet::PlanetType::None},
+		{"Mercury", Planet::PlanetType::Mercury},
+		{"Mars", Planet::PlanetType::Mars},
+		{"Ganymede", Planet::PlanetType::Ganymede}
+	};
+
+	const std::unordered_map<std::string, Planet::Hostility> hostilityTable
+	{
+		{"None", Planet::Hostility::None},
+		{"Low", Planet::Hostility::Low},
+		{"Medium", Planet::Hostility::Medium},
+		{"High", Planet::Hostility::High}
+	};
+
+
+	void parseElementValue(Planet::PlanetType& destination, const NAS2D::Xml::XmlElement* element)
+	{
+		destination = stringToEnum(planetTypeTable, element->getText());
+	}
+
+
+	void parseElementValue(Planet::Hostility& destination, const NAS2D::Xml::XmlElement* element)
+	{
+		destination = stringToEnum(hostilityTable, element->getText());
+	}
+}
+
+
+std::vector<Planet::Attributes> parsePlanetAttributes()
+{
+	const std::string rootElementName("Planets");
+	auto xmlDocument = openXmlFile("planets/PlanetAttributes.xml", rootElementName);
+
+	std::vector<Planet::Attributes> planetAttributes;
+
+	auto rootElement = xmlDocument.firstChildElement(rootElementName);
+	for (const auto* node = rootElement->iterateChildren(nullptr);
+		node != nullptr;
+		node = rootElement->iterateChildren(node))
+	{
+		std::string elementName("Planet");
+		if (node->value() != elementName)
+		{
+			throw std::runtime_error(xmlDocument.value() + " missing " + elementName + " tag");
+		}
+		planetAttributes.push_back(parsePlanet(node->toElement()));
+	}
+
+	return planetAttributes;
+}
+
+
+namespace 
+{
+	Planet::Attributes parsePlanet(const NAS2D::Xml::XmlElement* xmlNode)
+	{
+		Planet::Attributes attributes;
+
+		for (const auto* node = xmlNode->iterateChildren(nullptr);
+			node != nullptr;
+			node = xmlNode->iterateChildren(node))
+		{
+			const auto* element = node->toElement();
+
+			if (element->value() == "PlanetType")
+			{
+				parseElementValue(attributes.type, element);
+			}
+			else if (element->value() == "ImagePath")
+			{
+				::parseElementValue(attributes.imagePath, element);
+			}
```

Comment:
Hmm, yes, I see the discrepancy with the `::`. It seems you have two functions `parseElementValue`, defined in different namespaces. That could get confusing.

Reply:
Thank you for the review. My goal was to keep the function names the same. I reasoned it might be possible to use a table to match argument names to locations in the structure to remove the if / else logic statements to place each value. I thought doing this would be easier if function overloads / templates were used. It gets complicated because of the detailed knowledge required about the enum to parse it properly, so the local module will need to define overloads for which enums it wants to support parsing. If this seems silly, I'm OK if the enum parse functions have custom names...

I think it might be worth trying to limit enums when possible when parsing to simplify. We could just represent hostility as a scale of 1 to 10 using an int value instead of 'low' 'medium' 'high' for example which would cut out the custom parsing logic.

Reply:
The problem is the different attributes have different types, and selecting which function overload or template instantiation to call is a compile time thing. It's not compatible with a runtime loop. Trying to use a table would require some kind of compile time iteration.

One of the things I was thinking about was being able to map an enum type to the lookup table used to map strings to enum values. It is possible to use templates to map types to values:
https://gpfault.net/posts/mapping-types-to-values.txt.html

I was thinking if we used such a solution, we could update the `stringToEnum` function to not require the first parameter:
```cpp
template<typename T>
T stringToEnum(const std::unordered_map<std::string, T>& table, std::string value)
```
Instead, the table could be looked up internally using such a mapping template on the template parameter.

Taking that idea further, it could be combined with the `parseElementValue` function:
```cpp
template<typename T, std::enable_if_t<!std::is_enum<T>::value, bool> = true>
void parseElementValue(T& destination, const NAS2D::Xml::XmlElement* element)
```
You could potentially remove the `std::enable_if` guard, and instead use a `if constexpr` to select whether to call `parseElementValue` or `stringTo`. And you might as well, since there would no longer be an external `table` parameter to pass in, so it would be very easy to call in such a context.

Would then make it easy to use a return value, rather than an out parameter.

Of course, that all hinges on being able to map a type to a value. I gave it a bit of a try, though if you define the type mapping in a different namespace than it's used, that seems to cause issues.

----

```cpp
// Specialize this for each EnumType with a lookup table (not shown)
template <typename EnumType>
struct EnumValueName { static const std::unordered_map<std::string, EnumType> table; };

template<typename EnumType>
T stringToEnum(std::string value)
{
    const auto& table = EnumValueName<EnumType>::table;
    // Old implementation ...
}

template<typename T>
T parseElementValue(const NAS2D::Xml::XmlElement* element)
{
    try 
    {
        const auto value = element->getText();
        if constexpr (std::is_enum_v<T>)
        {
            return stringToEnum<T>(value);
        }
        else
        {
            return NAS2D::stringTo<T>(value);
        }
    }
    // Old implementation ...
}
```